feat: assets article (#700)
withastro/astro.build: src/content/blog/images.mdx

---
title: "Better Images in Astro"
description: "A new feature brings built-in image optimization to Astro. Coming in Astro 3.0, you can opt-in to try it yourself in today. Images in Astro have been completely redesigned with more features, faster performance, and a better end-user experience with automatic layout shift prevention."
publishDate: "June 7, 2023"
authors:
  - erika
  - fred
lang: "en"
coverImage: "/src/content/blog/_images/images/blog-hero-image-updates.jpg"
socialImage: "/src/content/blog/_images/images/og-image-image-updates.jpg"
---

import { Image } from "@astrojs/image/components"
import imageServiceImage from "./_images/images/service-1-image-updates.jpg"

Last year, we published the first Astro image integration ([@astrojs/image](https://docs.astro.build/en/guides/integrations-guide/image/)) alongside [Astro 1.0](https://astro.build/blog/astro-1). The goal was simple: **Make image optimization effortless in Astro.**

> Now, 1 year and 2.5 million downloads later, we're ready to share our new vision for working with images in Astro: completely redesigned with new features, faster performance, and a better end-user experience with [automatic layout shift (CLS) prevention](#automatic-layout-shift-prevention).

Reviewer: ```suggestion
Now, 1 year and 2.5 million downloads later, we're ready to share our new vision for working with images in Astro: completely redesigned with new features, faster performance, and a better end-user experience with:

- [Automatic layout shift (CLS) prevention](#automatic-layout-shift-prevention).
- [First-class Markdown, Markdoc & MDX support](#automatic-markdown--mdx-support).
- [Integration with third-party services](#integrating-with-third-party-image-services).
```

import { defineConfig } from "astro/config"

export default defineConfig({
	experimental: { assets: true },
})
```

In this post, we'll go over some of the new features and improvements that will make working with images in Astro even easier. If you prefer to jump right into the docs, read our [Assets](https://docs.astro.build/en/guides/assets/) guide for all of the details.

## How it works

From the start, we knew that we wanted an image API that felt as minimal as possible. We started with the original `<Image />` Component API from the `@astrojs/image` integration, and then built on it with a few new tweaks and changes.

```astro
---
// Import the <Image /> component
import { Image } from "astro:assets"
// Import a reference to the image itself
import myImage from "../assets/penguin.png"
---

<Image src={myImage} alt="A very cool penguin!" />
```

```html
<!-- Result: -->
<img
	src="/_astro/penguin.XXXXXX.webp"
	width="300"
	height="300"
	loading="lazy"
	decoding="async"
	alt="A very cool penguin!"
/>
```

But don't let the minimal API fool you:

- Astro optimizes your image using the modern [WebP](https://developers.google.com/speed/webp) file format (by default).

> - For accessibility, the required `alt` prop helps improve your site for screen-readers (and other tooling).

Reviewer: ```suggestion
- For accessibility, the required `alt` prop helps improve your site for screen-readers and other tooling.
```
No need to be parenthetical here! Embrace it!

**Astro's new `<Image />` component automatically protects your site from layout shift.** It works by inferring an explicit height and width for every image and including them in the generated `<img>` tag. With explicit dimensions, the image can't shift the page layout as it loads onto the page.

The output HTML is always a simple `<img>` element, so you can also apply additional CSS styling for more advanced responsive layouts like `fill` and `cover` without risking layout shift.

```astro
<Image
	src={myImage}
	alt="A very cool penguin!"
	style="height: 100%; width: 100%; object-fit: cover;"
/>
```

## Automatic Markdown & MDX support

One of the biggest improvements made over the original `@astrojs/image` integration is complete support for Markdown, Markdoc, and MDX. You can now reference images using a relative image path in your Markdown content or frontmatter, and Astro will automatically optimize them for you.

```md
![A starry night sky](../../assets/stars.png)
```

```html
<!-- Result: -->
<img
	src="/_astro/stars.XXXXXX.webp"
	width="300"
	height="300"
	loading="lazy"
	decoding="async"
	alt="A starry night sky"
/>
```

> That's it! Your `stars.png` image will be optimized automatically in the final HTML, as if you had used the `<Image>` component directly. You can also use the `<Image />` component directly in MDX if you prefer, but in most cases the default settings will be enough to produce an optimized file size with a great image quality.

Reviewer: ```suggestion
That's it! Your `stars.png` image will be optimized automatically in the final HTML. You can also use the `<Image />` component directly in MDX if you prefer for finer control over the image attributes, but in most cases the default settings will be enough to produce an optimized file size with a great image quality.
```
Tiny bit repetitive with "`<Image>` component directly", and I think it's more helpful to explain why you might Ithink you) want the image component.
